src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const array = [];
+  const stateClone = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateClone, action.extraData);
+        array.push({ ...stateClone });
+        break;
+
+      case 'removeProperties':
+        for (const character of action.keysToRemove) {
+          if (stateClone.hasOwnProperty(character)) {
+            delete stateClone[character];
+          }
+        }
+        array.push({ ...stateClone });
+        break;
+
+      case 'clear':
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
+        array.push({ ...stateClone });
```

Comment:
Remove redundant loop. In this task state is just copy
```suggestion
        stateClone = {}
        array.push({});
```